ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs

```diff
@@ -0,0 +1,695 @@
+    -- * Exported for ledger-db-backends-checker
+  , DbState (..)
+  , LMDBMK (..)
+  ) where
+
+import qualified Codec.Serialise as S (Serialise (..))
+import qualified Control.Concurrent.Class.MonadSTM.TVar as IOLike
+import           Control.Monad (unless, void, when)
+import qualified Control.Monad.Class.MonadSTM as IOLike
+import           Control.Monad.IO.Class (MonadIO (liftIO))
+import qualified Control.Tracer as Trace
+import           Data.Foldable (for_)
+import           Data.Functor (($>), (<&>))
+import           Data.Map (Map)
+import           Data.Map.Diff.Strict
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import qualified Data.Text as Strict
+import           GHC.Generics (Generic)
+
+import           Cardano.Slotting.Slot (SlotNo, WithOrigin (At))
+
+import           Ouroboros.Consensus.Ledger.Tables
+import qualified Ouroboros.Consensus.Storage.FS.API as FS
+import qualified Ouroboros.Consensus.Storage.FS.API.Types as FS
+import qualified Ouroboros.Consensus.Storage.LedgerDB.BackingStore as HD
+import qualified Ouroboros.Consensus.Storage.LedgerDB.BackingStore.LMDB.Bridge as Bridge
+import qualified Ouroboros.Consensus.Storage.LedgerDB.BackingStore.LMDB.TransactionHandle as TrH
+import           Ouroboros.Consensus.Util (foldlM', unComp2, (:..:) (..))
+import           Ouroboros.Consensus.Util.IOLike (Exception (..), IOLike,
+                     MonadCatch (..), MonadThrow (..), bracket)
+
+import qualified Database.LMDB.Raw as LMDB
+import qualified Database.LMDB.Simple as LMDB
+import qualified Database.LMDB.Simple.Cursor as LMDB.Cursor
+import qualified Database.LMDB.Simple.Extra as LMDB
+import qualified Database.LMDB.Simple.Internal as LMDB.Internal
+
+{-------------------------------------------------------------------------------
+ Backing Store interface
+-------------------------------------------------------------------------------}
+
+type LMDBBackingStoreInitialiser l m =
+  HD.BackingStoreInitialiser m
+    (LedgerTables l KeysMK)
+    (LedgerTables l ValuesMK)
+    (LedgerTables l DiffMK)
+
+type LMDBBackingStore l m =
+  HD.BackingStore m
+    (LedgerTables l KeysMK)
+    (LedgerTables l ValuesMK)
+    (LedgerTables l DiffMK)
+
+type LMDBValueHandle l m =
+  HD.BackingStoreValueHandle m
+    (LedgerTables l KeysMK)
+    (LedgerTables l ValuesMK)
+
+{-------------------------------------------------------------------------------
+ Tracing
+-------------------------------------------------------------------------------}
+
+data TraceDb
+  = TDBOpening
+  | TDBOpened      !FilePath
+  | TDBClosing     !FilePath
+  | TDBClosed      !FilePath
+  | TDBCopying     !FilePath !FilePath
+  | TDBCopied      !FilePath !FilePath
+  | TDBWrite       !(WithOrigin SlotNo) !SlotNo
+  | TDBValueHandle Int TraceValueHandle
+  | TDBTableOp     TraceTableOp
+  | TDBInitialisingFromLMDB !FS.FsPath
+  | TDBInitialisedFromLMDBD !FS.FsPath
+  deriving (Show, Eq)
```

Comment:
I'm a bit concerned that this might be too low level? Do we want to keep these traces around?

Reply:
Maybe consider reworking this a part of #4396?

Reply:
Yup, closing this thread